engine/common/follower/pending_tree/pending_tree.go

```diff
@@ -0,0 +1,198 @@
+//  4. Checks for exceeding byzantine threshold. Only one certified block per view is allowed.
+//  5. Finally, blocks with the lowest height from incoming batch that connect to the finalized state we will
+//     mark all descendants as connected, collect them and return as result of invocation.
+//
+// This function is designed to perform resolution of connected blocks(resolved block is the one that connects to the finalized state)
+// using incoming batch. Each block that was connected to the finalized state is reported once.
+// Expected errors during normal operations:
+//   - model.ByzantineThresholdExceededError - detected two certified blocks at the same view
+func (t *PendingTree) AddBlocks(certifiedBlocks []CertifiedBlock) ([]CertifiedBlock, error) {
+	// sort blocks by height, so we can identify if there are candidates for being connected to the finalized state.
+	slices.SortFunc(certifiedBlocks, func(lhs CertifiedBlock, rhs CertifiedBlock) bool {
+		return lhs.Height() < rhs.Height()
+	})
+
+	var allConnectedBlocks []CertifiedBlock
+	for _, block := range certifiedBlocks {
+		// skip blocks lower than finalized view
+		if block.View() <= t.forest.LowestLevel {
+			continue
+		}
+
+		iter := t.forest.GetVerticesAtLevel(block.View())
+		if iter.HasNext() {
+			v := iter.NextVertex().(*PendingBlockVertex)
+
+			if v.VertexID() == block.ID() {
+				// this vertex is already in tree, skip it
+				continue
+			} else {
+				return nil, model.ByzantineThresholdExceededError{Evidence: fmt.Sprintf(
+					"conflicting QCs at view %d: %v and %v",
+					block.View(), v.ID(), block.ID(),
+				)}
+			}
+		}
+
+		vertex, err := NewVertex(block, false)
+		if err != nil {
+			return nil, fmt.Errorf("could not create new vertex: %w", err)
+		}
+		err = t.forest.VerifyVertex(vertex)
+		if err != nil {
+			return nil, fmt.Errorf("failed to store certified block into the tree: %w", err)
+		}
+		t.forest.AddVertex(vertex)
+
+		if t.connectsToFinalizedBlock(block) {
+			connectedBlocks := t.updateAndCollectFork(vertex)
+			allConnectedBlocks = append(allConnectedBlocks, connectedBlocks...)
+		}
+	}
+
+	return allConnectedBlocks, nil
+}
+
+// connectsToFinalizedBlock checks if candidate block connects to the finalized state.
+func (t *PendingTree) connectsToFinalizedBlock(block CertifiedBlock) bool {
+	if block.Block.Header.ParentID == t.lastFinalizedID {
+		return true
+	} else if parentVertex, found := t.forest.GetVertex(block.Block.Header.ParentID); found {
+		return parentVertex.(*PendingBlockVertex).connectedToFinalized
+	} else {
+		return false
+	}
+}
+
+// FinalizeForkAtLevel takes last finalized block and prunes levels below the finalized view.
+// When a block is finalized we don't care for all blocks below it since they were already finalized.
+// No errors are expected during normal operation.
+func (t *PendingTree) FinalizeForkAtLevel(finalized *flow.Header) error {
```

Comment:
## ⚠️ 

I feel like there is an uncovered edge case here:
* I think we should allow the `PendingTree` to skip ahead. Assume we add some optimized logic where a node can provide a finalization proof many blocks into the future. 
   If we allow the `PendingTree` to skip ahead, we need to check whether the new finalized block has any children https://github.com/onflow/flow-go/blob/c0d5b849c354035d0f8ec86d99df3370ca797eea/engine/common/follower/pending_tree/pending_tree.go#L166 
* If we don't want to allow `PendingTree` to skip ahead (which I am also fine with), then we should enforce that `finalized` is already in the `PendingTree` and marked as `connected`

At the moment, we don't do either. Meaning, the `PendingTree` is externalizing a subtle but important consistency requirement (only known connected blocks are ever finalized).

Reply:
@AlexHentschel https://github.com/onflow/flow-go/pull/4003/commits/91a77b387fe0fb13d89946a0cdd2c92b04be1ee5